pkg/app/ops/deploymentchaincontroller/updater.go

```diff
@@ -0,0 +1,189 @@
+		logger:               lg,
+		nowFunc:              time.Now,
+	}
+}
+
+func (u *updater) IsDone() bool {
+	return u.done
+}
+
+func (u *updater) Run(ctx context.Context) error {
+	// In case all in chain applications' deployments are completed
+	// mark the updater as done and return immediately.
+	if u.isAllInChainDeploymentsCompleted() {
+		u.done = true
+		u.doneTimestamp = u.nowFunc()
+		return nil
+	}
+
+	// Check state of all deployment which existed from previous interval.
+	if len(u.deploymentsRef) != 0 {
+		for _, deploymentRef := range u.deploymentsRef {
+			// Ignore finished deployment.
+			if model.IsCompletedDeployment(deploymentRef.Status) {
+				continue
+			}
+			deployment, err := u.deploymentStore.GetDeployment(ctx, deploymentRef.DeploymentId)
+			if err != nil {
+				return err
+			}
+			// Update the deployment state in deployment chain model in case
+			// the deployment state is changed.
+			if deployment.Status != deploymentRef.Status {
+				if err = u.deploymentChainStore.UpdateDeploymentChain(ctx, u.deploymentChainID,
+					datastore.DeploymentChainNodeDeploymentStatusUpdater(
+						deployment.DeploymentChainBlockIndex,
+						deployment.Id,
+						deployment.Status,
+						deployment.StatusReason),
+				); err != nil {
+					return err
+				}
+				// Update updater's deploymentsRef.
+				deploymentRef.Status = deployment.Status
+				deploymentRef.StatusReason = deployment.StatusReason
+			}
+		}
+	}
+
+	// If not all applications in chain has its deployment ref
+	// fetch all missing deployments and link the deployment ref to them.
+	if len(u.deploymentsRef) != len(u.applicationsRef) {
+		deployments, err := u.listAllMissingDeployments(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Note: Only the deployment of the first block (the deployment which trigger this chain)
+		// is added at the time we created this deployment chain; thus all other deployments
+		// which be counted as "missng deployments" here are triggered and going to be added under
+		// PENDING state, since those deployments are of blocks which have to wait for its previous
+		// blocks completion. We don't need to care about updating those deployment ref status after
+		// they been added here in this interval.
+		for _, deployment := range deployments {
+			// Connect missing deployment to the chain.
+			if err := u.deploymentChainStore.UpdateDeploymentChain(
+				ctx,
+				u.deploymentChainID,
+				datastore.DeploymentChainAddDeploymentToBlock(deployment),
+			); err != nil {
+				return err
+			}
+			// Store deployment ref to the updater's deploymentsRef.
+			u.deploymentsRef[deployment.ApplicationId] = &model.ChainDeploymentRef{
+				DeploymentId: deployment.Id,
+				Status:       deployment.Status,
+				StatusReason: deployment.StatusReason,
+			}
+		}
+	}
+
+	return nil
+}
+
+func (u *updater) listAllMissingDeployments(ctx context.Context) ([]*model.Deployment, error) {
+	noDeploymentApps := make([]string, len(u.applicationsRef))
```

Comment:
Maybe length should be zero in case you append entities using `append`
```suggestion
	noDeploymentApps := make([]string, 0, len(u.applicationsRef))
```